boilerplates/react-union-boilerplate-liferay-ssr/liferay/ssr-filter/src/main/java/eu/reactunion/ssr/filter/ServerSideRenderingFilter.java

```diff
@@ -0,0 +1,120 @@
+import java.io.PrintWriter;
+
+@Component(
+        immediate = true,
+        property = {
+                "dispatcher=REQUEST",
+                "dispatcher=FORWARD",
+                "servlet-context-name=",
+                "servlet-filter-name=Server-side Rendering Filter",
+                "url-pattern=/*",
+                "after-filter=Cache Filter"
+        },
+        service = Filter.class
+)
+public class ServerSideRenderingFilter extends BaseFilter {
+    private static final Log _log = LogFactoryUtil.getLog(ServerSideRenderingFilter.class);
+
+    private static final String SSR_FLAG = "SSR";
+
+    private final RenderingServerApi renderingServerApi = Feign
+            .builder()
+            .target(RenderingServerApi.class, "http://localhost:3303");
+
+    private boolean renderingServerUp = true;
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+        super.init(filterConfig);
+        checkRenderingServerHealth();
+    }
+
+    @Override
+    protected void processFilter(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws Exception {
+        if (isDummyLayoutRequest(request)) {
+            super.processFilter(request, response, filterChain);
+            return;
+        }
+
+        BufferCacheServletResponse bufferCacheServletResponse = new BufferCacheServletResponse(response);
+        boolean requestAlreadyFlagged = isRequestFlagged(request);
+
+        if (renderingServerUp && !requestAlreadyFlagged) {
+            flagRequestForServerSideRendering(request);
+        } else {
+            flagRequestForClientSideRendering(request);
+        }
+
+        super.processFilter(request, bufferCacheServletResponse, filterChain);
+        String content = bufferCacheServletResponse.getString();
+        PrintWriter responseWriter = response.getWriter();
+
+        if (renderingServerUp && !requestAlreadyFlagged && isResponseRenderable(response)) {
+            try {
+                responseWriter.write(renderingServerApi.render(content));
+            } catch (Exception ex) {
+                _log.error("Rendering server failed to render.", ex);
+                checkRenderingServerHealth();
+                processFilter(request, response, filterChain);
+            }
+        } else {
+            responseWriter.write(content);
+        }
+    }
+
+    private boolean isResponseRenderable(HttpServletResponse response) {
+        return response.getContentType() != null && response.getContentType().contains("text/html");
+    }
+
+    private boolean isDummyLayoutRequest(HttpServletRequest request) {
+        return request.getRequestURI() != null && request.getRequestURI().startsWith("/c/portal/layout");
+    }
+
+    private void checkRenderingServerHealth() {
+        try {
+            renderingServerApi.checkHealth();
+            renderingServerUp = true;
+        } catch (Exception ex) {
+            renderingServerUp = false;
```

Comment:
How can you put it back online. There is no way how to do it. What about scheduled health check with configurable periodicity or default 5 min periodicity. So the worst case will be, that after 5 minutes it could be marked up again.